src/middlewares/restrict-user-details.js

```diff
@@ -0,0 +1,102 @@
+        
+        ctx.response.body = fieldsRestricter.restrictFields(ctx.response.body, modelName, action, user);
+      }
+    } catch (error) {
+      console.error(error);
+    }
+  };
+};
+
+class FieldsRestricter {
+  constructor(metadata, config) {
+    this.metadata = metadata;
+    this.config = config;
+  }
+
+  findAndRestrictUserDetails(item, modelName) {
+    if (modelName === "user") {
+      return this.restrictNonWrappedItem(item, this.config.allowedUserFields);
+    }
+    if (modelName === "plugin::users-permissions.user") {
+      return this.restrictWrappedItem(item, this.config.allowedUserFields);
+    } 
+    if (modelName === "api::profile.profile") {
+      return this.restrictWrappedItem(item, this.config.allowedProfileFields);
+    }
+    if (!this.metadata.has(modelName)) {
+      return item;
+    }
+    const model = this.metadata.get(modelName);
+    if (model && item?.data) {
+      if (Array.isArray(item.data)) {
+        item.data.forEach((element) => {
+          this.reviewModelAttributes(element.attributes, model);
+        });
+      } else {
+        this.reviewModelAttributes(item.data.attributes, model);
+      }
+    }
+    return item;
+  }
+
+  reviewModelAttributes(item, model) {
+    for (let attribute in model.attributes) {
+      const attributeData = model.attributes[attribute];
+      if (attributeData.type === "relation" && item[attribute]) {
+        item[attribute] = this.findAndRestrictUserDetails(item[attribute], attributeData.target);
+      }
+    }
+  }
+
+  restrictWrappedItem(item, allowedFields) {
+    const restrictedItem = {data : {id: item.data.id, attributes: {}}};
+    for (const field of allowedFields) {
+      if (field === "profile" && item.data.attributes[field]?.data) {
+        restrictedItem.data.attributes[field] = this.restrictWrappedItem(item.data.attributes[field], this.config.allowedProfileFields);
+      } else if (item.data.attributes[field]) {
+        restrictedItem.data.attributes[field] = item.data.attributes[field];
+      }
+    }
+    return restrictedItem;
+  }
+
+  restrictNonWrappedItem(item, allowedFields) {
+    const restrictedItem = {};
+    for (const field of allowedFields) {
+      if (field === "profile" && item[field]) {
+        restrictedItem[field] = this.restrictNonWrappedItem(item[field], this.config.allowedProfileFields);
+      } else if (item[field]) {
+        restrictedItem[field] = item[field];
+      }
+    }
+    return restrictedItem;
+  }
+
+  restrictFields(response, modelName, action, user) {
+    if (modelName === "user" && action === "me") {
+      return response;
+    }
+    if (modelName === "user" && action === 'findOne' && response?.id === user.id) {
+      return response;
+    }
+    if (modelName === "api::profile.profile" && action === 'findOne' && response?.data?.attributes?.user?.data?.id === user.id) {
+      return response;
+    }
+    return this.findAndRestrictUserDetails(response, modelName);
```

Comment:
Main algorithm here: take the response and introspect all attributes, if any attribute is a relation, we take that object and introspect its attributes. When relation points to `user` or `profile`, we replace that object with shrunk version, including only allowed fields.